### PR TITLE
WIP: Add logic to support local dev environment

### PR DESCRIPTION
**Pull request by sigalsax**
We are pursuing an effort to build a local dev environment so that instead of pushing to a remote K8s cluster, developers can deploy the cluster locally with Docker desktop
- Add dev checks to block from unnecessary code from running
- Create separate bootstrap.env file for local development

TODO:
- [ ] Add documentation for deploying Conjur for local development
- [ ] Provide automated scripts for easy deployment with Secrets Provider

### CODE REVIEW

File: 2_prepare_docker_images.sh [hunk 1]
```diff
@@ -26,11 +26,12 @@ prepare_conjur_appliance_image() {
   conjur_appliance_image=$(platform_image conjur-appliance)
 
   # Try to pull the image if we can
-  docker pull $CONJUR_APPLIANCE_IMAGE || true
-
+  if [ ! is_dev_env ]; then
```
Comment (abrahamko) [inline 2_prepare_docker_images.sh:29]: Why we don't need this in dev env?

File: 2_prepare_docker_images.sh [hunk 2]
```diff
@@ -44,7 +45,8 @@ prepare_conjur_cli_image() {
   cli_app_image=$(platform_image conjur-cli)
   docker tag conjur-cli:$CONJUR_NAMESPACE_NAME $cli_app_image
 
-  if ! is_minienv; then
+  if [ ! is_minienv ] && [ ! is_dev_env ]; then
+    echo "HI"
```
Comment (abrahamko) [inline 2_prepare_docker_images.sh:49]: 👋 ?

File: 3_deploy_conjur_master_cluster.sh [hunk 3]
```diff
@@ -6,7 +6,9 @@ set -eo pipefail
 main() {
   set_namespace $CONJUR_NAMESPACE_NAME
 
-  docker_login
+  if ! is_dev_env; then
```
Comment (abrahamko) [inline 3_deploy_conjur_master_cluster.sh:9]:
```suggestion
  if [ ! is_dev_env ]; then
```

File: 3_deploy_conjur_master_cluster.sh [hunk 4]
```diff
@@ -6,7 +6,9 @@ set -eo pipefail
...
+    docker_login
```
Comment (abrahamko) [inline 3_deploy_conjur_master_cluster.sh:10]: Is this really not needed? What about first dev env preparation?
Let's think about the next developer who joins that might use this to prepare his env.

File: 3_deploy_conjur_master_cluster.sh [hunk 5]
```diff
@@ -62,6 +64,9 @@ deploy_conjur_cli() {
   announce "Deploying Conjur CLI pod."
 
   cli_app_image=$(platform_image conjur-cli)
+
+  $cli delete --ignore-not-found deployment conjur-cli
```
Comment (abrahamko) [inline 3_deploy_conjur_master_cluster.sh:68]: Add a comment why you run delete here

File: 4_deploy_conjur_followers.sh [hunk 6]
```diff
@@ -48,6 +48,9 @@ docker_login() {
 deploy_conjur_followers() {
   announce "Deploying Conjur Follower pods."
 
+  $cli delete --ignore-not-found deployment conjur-follower
```
Comment (abrahamko) [inline 4_deploy_conjur_followers.sh:51]: Add a comment why you run delete here

File: 4_deploy_conjur_followers.sh [hunk 7]
```diff
@@ -93,6 +96,9 @@ deploy_conjur_followers() {
 add_server_certificate_to_configmap() {
   SERVER_CERTIFICATE="./server_certificate.cert"
   ./_save_server_cert.sh $SERVER_CERTIFICATE
+
+  $cli delete --ignore-not-found configmap server-certificate
```
Comment (abrahamko) [inline 4_deploy_conjur_followers.sh:100]: Add a comment why you run delete here

File: 5_create_load_balancer.sh [hunk 8]
```diff
@@ -15,6 +15,9 @@ deploy_load_balancer() {
 
   set_namespace $CONJUR_NAMESPACE_NAME
 
+  $cli delete --ignore-not-found service conjur-master-ext
```
Comment (abrahamko) [inline 5_create_load_balancer.sh:18]: Add a comment why you run delete here

File: dev-bootstrap.env [hunk 9]
```diff
@@ -57,6 +57,14 @@ export CONJUR_ADMIN_PASSWORD=ADmin123!!!!
 # export DOCKER_REGISTRY_URL=<registry-domain>
 # export DOCKER_REGISTRY_PATH=<registry-domain>/<additional-pathing>
 
+#######
+# LOCAL DEV CONFIG USING DOCKER DESKTOP (uncomment all lines if using this configuration)
```
Comment (abrahamko) [inline dev-bootstrap.env:61]: Isn't there a better way to configure this besides commenting out these lines?

File: utils.sh [hunk 10]
```diff
@@ -6,8 +6,9 @@ DEPLOY_MASTER_CLUSTER="${DEPLOY_MASTER_CLUSTER:-false}"
 FOLLOWER_USE_VOLUMES="${FOLLOWER_USE_VOLUMES:-false}"
 
 MINI_ENV="${MINI_ENV:-false}"
+DEV="${DEV:-false}"
 
-if [[ "$MINI_ENV" == "true" ]]; then
+if [ "$MINI_ENV" == "true" ] || [ "$DEV" == "true" ]; then
```
Comment (abrahamko) [inline utils.sh:11]: Why not using `is_dev_env`, like anywhere else?

File: 0_check_dependencies.sh [hunk 11]
```diff
@@ -7,11 +7,11 @@ check_env_var "CONJUR_APPLIANCE_IMAGE"
 check_env_var "CONJUR_NAMESPACE_NAME"
 check_env_var "AUTHENTICATOR_ID"
 
-if [ ! is_minienv ]; then
+if [ ! is_minienv ] || [ ! is_dev_env ]; then
   check_env_var "DOCKER_REGISTRY_PATH"
 fi
 
-if [ "${PLATFORM}" = "kubernetes" ] && [ ! is_minienv ]; then
+if [ "${PLATFORM}" = "kubernetes" ] && [ ! is_minienv ] || [ ! is_dev_env ]; then
```
Comment (orenbm) [inline 0_check_dependencies.sh:14]: can we add brackets to emphasize the order? I _think_ it can be done with `[ "${PLATFORM}" = "kubernetes" ] && ([ ! is_minienv ] || [ ! is_dev_env ])`

File: 1_prepare_conjur_namespace.sh [hunk 12]
```diff
@@ -62,7 +62,7 @@ create_cluster_role() {
 }
 
 create_cluster_role_binding() {
```
Comment (orenbm) [inline 1_prepare_conjur_namespace.sh:64]: i see that you use a RoleBinding instead of a ClusterRoleBinding. Can you please split this into a separate commit and explain why you are performing this change? It will be interesting to know why RoleBinding is better.

Also - please change the method name while you're at it.

File: 2_prepare_docker_images.sh [hunk 13]
```diff
@@ -26,11 +26,12 @@ prepare_conjur_appliance_image() {
   conjur_appliance_image=$(platform_image conjur-appliance)
 
   # Try to pull the image if we can
-  docker pull $CONJUR_APPLIANCE_IMAGE || true
-
+  if [ ! is_dev_env ]; then
+    docker pull $CONJUR_APPLIANCE_IMAGE || true
+  fi
   docker tag $CONJUR_APPLIANCE_IMAGE $conjur_appliance_image
```
Comment (orenbm) [inline 2_prepare_docker_images.sh:32]: can you add a blank line before this?

File: 2_prepare_docker_images.sh (hunk 2 above)
Comment (orenbm) [reply 2_prepare_docker_images.sh:49]: HI @sigalsax :)

File: 2_prepare_docker_images.sh [hunk 14]
```diff
@@ -54,10 +56,10 @@ prepare_seed_fetcher_image() {
 
   docker pull $SEEDFETCHER_IMAGE
 
-  seedfetcher_image=$(platform_image seed-fetcher)
+  seedfetcher_image=seed-fetcher:$CONJUR_NAMESPACE_NAME
```
Comment (orenbm) [inline 2_prepare_docker_images.sh:59]: can you please elaborate on this change?

File: dev-bootstrap.env [hunk 15]
```diff
@@ -57,6 +57,14 @@ export CONJUR_ADMIN_PASSWORD=ADmin123!!!!
...
+#######
+# export PLATFORM=kubernetes
+# export DEV=true
+# export DEPLOY_MASTER_CLUSTER="true"
+# export STOP_RUNNING_ENV=false
```
Comment (orenbm) [inline dev-bootstrap.env:66]: why don't we need to stop the running env? are we assuming that the user ran something in k8s before running this script?

File: utils.sh [hunk 16]
```diff
@@ -160,6 +161,14 @@ is_minienv() {
   fi
 }
 
+is_dev_env() {
```
Comment (orenbm) [inline utils.sh:164]: do we need a method for this? can't we write `is_dev_env=${ENV-false}`?

File: 2_prepare_docker_images.sh (hunk 2 above)
Comment (sigalsax) [reply 2_prepare_docker_images.sh:49]: LOL darnnit -.-"

**Comment by sigalsax** [comment]
closing due to branch messiness. New PR: https://github.com/cyberark/kubernetes-conjur-deploy/pull/132

File: 1_prepare_conjur_namespace.sh (hunk 12 above)
Comment (orenbm) [reply 1_prepare_conjur_namespace.sh:64]: oh i see now that it was always a rolebinding. that's odd. can you please just change the method name?

Also - i see that we have a RoleBinding and a ClusterRole. does that make sense? shouldn't it be a Role? (i know it's not part of your PR but you researched that area lately)

File: 3_deploy_conjur_master_cluster.sh (hunk 4 above)
Comment (sigalsax) [reply 3_deploy_conjur_master_cluster.sh:10]: you are right. deleted the if statement in both `3_` and `4_`

File: dev-bootstrap.env (hunk 9 above)
Comment (sigalsax) [reply dev-bootstrap.env:61]: probably but source bootstrap.env is the flow for this project so considering another direction will probably be in a separate PR
